File: 300/309 - Best Time to Buy and Sell Stock with Cooldown.c
#include <limits.h>

#define max(a, b) ((a) > (b) ? (a) : (b))

int maxProfit(int* prices, int pricesSize) {
    int sell = 0, buy = INT_MIN, p_sell = 0, p_buy = 0, i, price;
    
    for (i = 0; i < pricesSize; ++i) {
        price = prices[i];
        
        p_buy = buy;
        buy = max(p_sell - price, buy);
        p_sell = sell;
        sell = max(p_buy + price, sell);
    }
    
    return sell;
}

File: csharp/309.best-time-to-buy-and-sell-stock-with-cooldown.cs
/*
 * @lc app=leetcode id=309 lang=csharp
 *
 * [309] Best Time to Buy and Sell Stock with Cooldown
 */
public class Solution {
    public int MaxProfit(int[] prices) {
        int buy = int.MinValue, pre_buy = int.MinValue, sell = 0, pre_sell = 0;
        foreach (int price in prices) {
            pre_buy = buy;
            buy = Math.Max(pre_sell - price, pre_buy);
            pre_sell = sell;
            sell = Math.Max(pre_sell, pre_buy + price);
        }
        return sell;
    }
}

